File: scala/problem2.scala
object Main {
  def main(args: Array[String]) {
    val answer = fibonacciNumbers(from = 1, max = 4000000).filter(n => n % 2 == 0).sum
    println(answer)
  }

  def fibonacciNumbers(from: Int, max: Int): List[Int] = {
    fibonacci(from) match {
      case n if n < max => fibonacci(from) :: fibonacciNumbers(from + 1, max)
      case _ => List.empty[Int]
    }
  }

  def fibonacci(n: Int): Int = {
    n match {
      case 1 => 1
      case 2 => 2
      case n => fibonacci(n-1) + fibonacci(n-2)
    }
  }
}

File: rust/problem2.rs
fn main() {
    fn fibonacci_numbers(n: u32, limit: u32) -> Vec<u32> {
        let fib = fibonacci(n);

        match fib {
          f if f < limit => fib :: fibonacci_numbers(n+1, limit),
          _ => vec![]
        }

        //if fibonacci_n <= 4000000 {
        //    if fibonacci_n % 2 == 0 {
        //        let mut vec = vec![fibonacci_n];
        //        vec.append(&mut fibonacci_numbers(n+1));
        //        return vec;
        //    } else {
        //        return fibonacci_numbers(n+1);
        //    }
        //} else {
        //    let v : Vec<u32> = Vec::new();
        //    return v;
        //}
    }

    //for n in fibonacci_numbers(1) {
    //    println!("{}", n);
    //}

    let sum = fibonacci_numbers(1).iter().fold(0, |sum, v| sum + v );
    println!("sum: {}", sum);
}

fn fibonacci(n: u32) -> u32 {
    match n {
      n if n <= 2 => n,
      n => fibonacci(n-1) + fibonacci(n-2)
    }
}
